pkg/pool-weighted/contracts/managed/ManagedPool.sol

```diff
@@ -120,13 +120,26 @@ contract ManagedPool is ManagedPoolSettings {
         SwapRequest memory request,
         uint256 balanceTokenIn,
         uint256 balanceTokenOut
-    ) internal view override returns (uint256) {
+    ) internal override returns (uint256) {
         bytes32 poolState = _getPoolState();
         _require(ManagedPoolStorageLib.getSwapsEnabled(poolState), Errors.SWAPS_DISABLED);
 
         // solhint-disable no-empty-blocks
         if (request.tokenOut == IERC20(this)) {
-            // Do a joinSwap
+            // Check allowlist for LPs, if applicable
+            _require(isAllowedAddress(request.from), Errors.ADDRESS_NOT_ALLOWLISTED);
+
+            // balanceTokenOut is the amount of BPT held by the Pool in the Vault.
+            // Subtracting this from the total supply gives us the virtual supply.
+            uint256 virtualSupply = totalSupply() - balanceTokenOut;
+
+            // The AUM fee calculation is based on inflating the Pool's BPT supply by a target rate.
+            // We then must collect AUM fees whenever joining or exiting the pool to ensure that LPs only pay AUM fees
+            // for the period during which they are an LP within the pool: otherwise an LP could shift their share of
+            // the AUM fees onto the remaining LPs in the pool by exiting before they were paid.
+            uint256 actualSupply = virtualSupply + _collectAumManagementFees(virtualSupply);
```

Comment:
This seems rather odd - why do this here and not in `_onJoinSwap`?

Reply:
I guess ultimately it doesn't matter too much, we can always rearrange these later on.

Reply:
It feels more right to me to do handling of fees and permissioning here and then have  `_onJoinSwap` be a pure "I'm getting A so you get B" function.